### PR TITLE
Remove mmap-based index header reader

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR removes the mmap-based index header reader from the store gateway, and makes the mmap-less index header reader the default.

It also removes the `fileutil` package, which contained only mmap-related functionality that is no longer used anywhere.

Only thing I was unsure about here was whether or not we want to remove the `-blocks-storage.bucket-store.index-header.stream-reader-enabled` flag immediately, or keep it (but ignore it) until the major release after the next major release.

#### Which issue(s) this PR fixes or relates to

https://github.com/grafana/mimir/issues/3465

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/indexheader/header.go
Comment (pracucci) [inline pkg/storegateway/indexheader/header.go:48]: You should remove this too.

File: pkg/storegateway/indexheader/header.go
Comment (pracucci) [inline pkg/storegateway/indexheader/header.go:49]: This shouldn't be experimental anymore (but advanced).

**Review by pracucci — APPROVED** [review]
LGTM (modulo few comments)! Also, before merging, can you also remove the flag reference from `docs/sources/mimir/operators-guide/configure/about-versioning.md`?

> Only thing I was unsure about here was whether or not we want to remove the -blocks-storage.bucket-store.index-header.stream-reader-enabled flag immediately, or keep it (but ignore it) until the major release after the next major release.

It was experimental, so it's OK to remove it without deprecating it first.

File: pkg/storegateway/indexheader/header.go
Comment (charleskorn) [reply pkg/storegateway/indexheader/header.go]: Done in e18a97f.

File: pkg/storegateway/indexheader/header.go
Comment (charleskorn) [reply pkg/storegateway/indexheader/header.go]: Done in e18a97f.

**Comment by charleskorn** [comment]
> Also, before merging, can you also remove the flag reference from `docs/sources/mimir/operators-guide/configure/about-versioning.md`?

Done.

I've also renamed the `stream-reader-max-idle-file-handles` flag to `max-idle-file-handles` given there's now no other kind of reader.

**Review by pracucci — APPROVED** [review]
Thanks!

**Pull request merged**
